### PR TITLE
[5.1] DB Query Builder Join Clause 'where in'

### PR DESCRIPTION
**Pull request by mightydes**
Added Join Clause methods:
- `whereIn`
- `whereNotIn`
- `orWhereIn`
- `orWhereNotIn`

[Issue discussion](https://github.com/laravel/framework/issues/4412)
[previously](https://github.com/laravel/framework/pull/8884)

### CODE REVIEW

File: src/Illuminate/Database/Query/JoinClause.php [hunk 1]
```diff
@@ -57,12 +57,15 @@ public function __construct($type, $table)
...
     public function on($first, $operator, $second, $boolean = 'and', $where = false)
     {
-        $this->clauses[] = compact('first', 'operator', 'second', 'boolean', 'where');
+        $isIn = $where && ($operator === 'in' || $operator === 'not in') && is_array($second);
 
         if ($where) {
             $this->bindings[] = $second;
         }
 
+        !$isIn ?: $second = count($second);
```
Comment (taylorotwell) [inline src/Illuminate/Database/Query/JoinClause.php]: This is crazy looking. Please change to something more immediately readable.

**Comment by taylorotwell** [comment]
Code is not very readable. Has no tests. Feel free to try again.

File: src/Illuminate/Database/Query/JoinClause.php (hunk 1 above)
Comment (mightydes) [inline src/Illuminate/Database/Query/JoinClause.php]: Why are you so hard on Elvis operator? ))

**Comment by mightydes** [comment]
@taylorotwell
Which branch this time?)
And what you suggest for test class name?

**Comment by Shkeats** [comment]
Really keen on this as a feature. Thanks for the effort @mightydes, keep it up.

**Pull request closed without merging**